src/detect-dns-response.c

```diff
@@ -0,0 +1,212 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+/**
+ * \ingroup dnslayer
+ *
+ * @{
+ */
+
+
+/**
+ * \file
+ *
+ * \author Stian Bergseth <stianb@mnemonic.no>
+ */
+
+#include "suricata-common.h"
+#include "threads.h"
+#include "debug.h"
+#include "decode.h"
+#include "detect.h"
+
+#include "detect-parse.h"
+#include "detect-engine.h"
+#include "detect-engine-mpm.h"
+#include "detect-content.h"
+#include "detect-pcre.h"
+
+#include "flow.h"
+#include "flow-util.h"
+#include "flow-var.h"
+
+#include "util-debug.h"
+#include "util-unittest.h"
+#include "util-spm.h"
+#include "util-print.h"
+
+#include "stream-tcp.h"
+
+#include "app-layer.h"
+#include "app-layer-dns-common.h"
+#include "detect-dns-response.h"
+#include "detect-engine-dns.h"
+
+#include "util-unittest-helper.h"
+
+static int DetectDnsResponseSetup (DetectEngineCtx *, Signature *, const char *);
+static void DetectDnsResponseRegisterTests(void);
+static int g_dns_response_buffer_id = 0;
+
+void DetectDnsResponseRegister (void)
+{
+    sigmatch_table[DETECT_AL_DNS_RESPONSE].name = "dns_response";
+    sigmatch_table[DETECT_AL_DNS_RESPONSE].desc = "content modifier to match DNS responses for A records";
+    sigmatch_table[DETECT_AL_DNS_RESPONSE].Match = NULL;
+    sigmatch_table[DETECT_AL_DNS_RESPONSE].Setup = DetectDnsResponseSetup;
+    sigmatch_table[DETECT_AL_DNS_RESPONSE].Free  = NULL;
+    sigmatch_table[DETECT_AL_DNS_RESPONSE].RegisterTests = DetectDnsResponseRegisterTests;
+
+    sigmatch_table[DETECT_AL_DNS_RESPONSE].flags |= SIGMATCH_NOOPT;
+
+    DetectAppLayerMpmRegister("dns_response", SIG_FLAG_TOCLIENT, 2,
+        PrefilterTxDnsResponseRegister);
+
+    DetectAppLayerInspectEngineRegister("dns_response",
+            ALPROTO_DNS, SIG_FLAG_TOCLIENT, 1,
+            DetectEngineInspectDnsResponse);
+
+    g_dns_response_buffer_id = DetectBufferTypeGetByName("dns_response");
+
+    DetectBufferTypeSetDescriptionByName("dns_response",
+            "dns response");
+}
+
+/**
+ * \brief this function setups the dns_response modifier keyword used in the rule
+ *
+ * \param de_ctx   Pointer to the Detection Engine Context
+ * \param s        Pointer to the Signature to which the current keyword belongs
+ * \param str      Should hold an empty string always
+ *
+ * \retval  0 On success
+ */
+
+static int DetectDnsResponseSetup(DetectEngineCtx *de_ctx, Signature *s, const char *str)
+{
+    s->init_data->list = g_dns_response_buffer_id;
+    s->alproto = ALPROTO_DNS;
+    return 0;
+}
+
+
+
+#ifdef UNITTESTS
+
+/** \test simple dns response match on A record */
+static int DetectDnsResponseTest01(void)
```

Comment:
can you convert this to fully using the FAIL_IF style macro's only? No conditions, just a linear path of checks until PASS.